src/limit.cpp

```diff
@@ -0,0 +1,295 @@
+//
+//            LimitProperties Class
+//
+//----------------------------------------------
+LimitProperties::LimitProperties()
+{
+	_type = LimitType::None;
+	_value = 0;
+	_auto_release = true;
+};
+
+LimitProperties::~LimitProperties()
+{
+};
+
+void LimitProperties::init() 
+{
+	_type = LimitType::None;
+	_value = 0;
+	_auto_release = true;
+};
+
+LimitType LimitProperties::getType() {
+	return _type;
+};
+
+bool LimitProperties::setType(LimitType type)
+{
+	_type = type;
+	return true;
+};
+
+uint32_t  LimitProperties::getValue() {
+	return _value;
+};
+
+bool LimitProperties::setValue(uint32_t value)
+{
+	_value = value;
+	return true;
+};
+
+bool LimitProperties::getAutoRelease() {
+	return _auto_release;
+};
+
+bool LimitProperties::deserialize(JsonObject &obj)
+{
+	if(obj.containsKey("type")) {
+		_type.fromString(obj["type"]);
+  	}
+	if(obj.containsKey("value")) {
+		_value = obj["value"];
+  	}
+	if(obj.containsKey("auto_release")) {
+		_auto_release = obj["auto_release"];
+  	}
+	return _type > 0 && _value > 0;
+
+};
+
+bool LimitProperties::serialize(JsonObject &obj)
+{
+
+	obj["type"] = _type.toString();
+	obj["value"] = _value;
+	obj["auto_release"] = _auto_release;
+	return true;
+};
+
+// ---------------------------------------------
+//
+//            Limit Class
+//
+//----------------------------------------------
+
+//global instance
+Limit limit;
+
+Limit::Limit() : Limit::Task() {
+	_version = 0;
+	_limit_properties.init();
+	StaticJsonDocument<32> doc;
+	doc["limit_version"] = ++_version;
+	event_send(doc);
+};
+
+Limit::~Limit() {
+	_evse -> release(EvseClient_OpenEVSE_Limit);
+};
+
+void Limit::setup() {
+
+};
+
+void Limit::begin(EvseManager &evse) {
+	// todo get saved default limit
+	DBUGLN("Starting Limit task");
+	this -> _evse    = &evse;
+	MicroTask.startTask(this);
+};
+
+unsigned long Limit::loop(MicroTasks::WakeReason reason) {
+
+
+	if (hasLimit()) {
+		LimitType type = _limit_properties.getType();
+		uint32_t value = _limit_properties.getValue();
+		bool auto_release = _limit_properties.getAutoRelease();
+
+		if (_evse->isCharging() ) {
+			_has_vehicle = true;
+			bool limit_reached = false;
+			switch (type) {
+				case LimitType::Time:
+					limit_reached = limitTime(value);
+					break;
+				case LimitType::Energy:
+					limit_reached = limitEnergy(value);
+					break;
+				case LimitType::Soc:
+					limit_reached = limitSoc(value);
+					break;
+				case LimitType::Range:
+					limit_reached = limitRange(value);
+					break;
+			}
+			if (limit_reached) {
+				// Limit reached, disabling EVSE
+				if (_evse->getClaimProperties(EvseClient_OpenEVSE_Limit).getState() == EvseState::None) {
+					DBUGLN("Limit as expired, disable evse");
+					EvseProperties props;
+					props.setState(EvseState::Disabled);
+					props.setAutoRelease(true);
+					_evse->claim(EvseClient_OpenEVSE_Limit, EvseManager_Priority_Limit, props);
+				}
+			}
+		}
+
+		else if ( _has_vehicle && !_evse->isVehicleConnected()) {
```

Comment:
There is an `onSessionComplete` complete event you should use for this for consistancy

Reply:
ah good catch, far better.